apidef/oas/upstream.go

```diff
@@ -181,6 +222,122 @@ func (u *Upstream) loadBalancingExtractTo(api *apidef.APIDefinition) {
+	// If InsecureSkipVerify is true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
+	// In this mode, TLS is susceptible to machine-in-the-middle attacks unless custom verification is used.
+	// This should be used only for testing or in combination with VerifyConnection or VerifyPeerCertificate.
+	//
+	// Tyk classic API definition: `proxy.transport.ssl_insecure_skip_verify`
+	InsecureSkipVerify bool `bson:"insecureSkipVerify,omitempty" json:"insecureSkipVerify,omitempty"`
+
+	// Ciphers is a list of SSL ciphers to be used. If unset, the default ciphers will be used.
+	//
+	// Tyk classic API definition: `proxy.transport.ssl_ciphers`
+	Ciphers []string `bson:"ciphers,omitempty" json:"ciphers,omitempty"`
+
+	// MinVersion is the minimum SSL/TLS version that is acceptable.
+	// Tyk classic API definition: `proxy.transport.ssl_min_version`
+	MinVersion string `bson:"minVersion,omitempty" json:"minVersion,omitempty"`
+
+	// MaxVersion is the maximum SSL/TLS version that is acceptable.
+	MaxVersion string `bson:"maxVersion,omitempty" json:"maxVersion,omitempty"`
+
+	// ForceCommonNameCheck forces the validation of the hostname against the certificate Common Name.
+	//
+	// Tyk classic API definition: `proxy.transport.ssl_force_common_name_check`
+	ForceCommonNameCheck bool `bson:"forceCommonNameCheck,omitempty" json:"forceCommonNameCheck,omitempty"`
+}
+
+// Fill fills *TLSTransport from apidef.ServiceDiscoveryConfiguration.
+func (t *TLSTransport) Fill(api apidef.APIDefinition) {
+	t.ForceCommonNameCheck = api.Proxy.Transport.SSLForceCommonNameCheck
+	t.Ciphers = api.Proxy.Transport.SSLCipherSuites
+	t.MaxVersion = tlsVersionToString(api.Proxy.Transport.SSLMaxVersion)
+	t.MinVersion = tlsVersionToString(api.Proxy.Transport.SSLMinVersion)
+	t.InsecureSkipVerify = api.Proxy.Transport.SSLInsecureSkipVerify
+}
+
+// ExtractTo extracts *TLSTransport into *apidef.ServiceDiscoveryConfiguration.
+func (t *TLSTransport) ExtractTo(api *apidef.APIDefinition) {
+	api.Proxy.Transport.SSLForceCommonNameCheck = t.ForceCommonNameCheck
+	api.Proxy.Transport.SSLCipherSuites = t.Ciphers
+	api.Proxy.Transport.SSLMaxVersion = tlsVersionFromString(t.MaxVersion)
+	api.Proxy.Transport.SSLMinVersion = tlsVersionFromString(t.MinVersion)
+	api.Proxy.Transport.SSLInsecureSkipVerify = t.InsecureSkipVerify
+}
+
+// tlsVersionFromString converts v in the form of 1.2/1.3 to the version int
+func tlsVersionFromString(v string) uint16 {
+	switch v {
+	case "1.0":
+		return tls.VersionTLS10
+	case "1.1":
+		return tls.VersionTLS11
+	case "1.2":
+		return tls.VersionTLS12
+	case "1.3":
+		return tls.VersionTLS13
+	default:
+		return 0
+	}
+}
+
+// tlsVersionFromString converts v from version into to the form 1.0/1.1
+func tlsVersionToString(v uint16) string {
+	switch v {
+	case tls.VersionTLS10:
+		return "1.0"
+	case tls.VersionTLS11:
+		return "1.1"
```

Comment:
This needs to be in internal/crypto or httputil (whereever more tls code lives, iirc crypto)